classes/edge.bbclass

```diff
@@ -0,0 +1,89 @@
+#       install -d ${D}${EDGE_BIN}
+#       install -m 0755 ${B}/${GO_BUILD_BINDIR}/edge-proxy ${D}${EDGE_BIN}/
+#
+#
+  
+#-----------------------------------------------------------------------------------------------------#
+#                                                  _   _                                              #
+#                                      _ __   __ _| |_| |__  ___                                      #
+#                                     | '_ \ / _` | __| '_ \/ __|                                     #
+#                                     | |_) | (_| | |_| | | \__ \                                     #
+#                                     | .__/ \__,_|\__|_| |_|___/                                     #
+#                                     |_|                                                             #
+#-----------------------------------------------------------------------------------------------------#
+#/usr/bin
+EDGE_BIN = "${bindir}"
+
+#/usr/lib/edge
+EDGE_LIB = "${libdir}/edge"
+
+#/etc/edge
+EDGE_ETC = "${sysconfdir}/edge"
+
+#/var/rootdirs/userdata
+EDGE_DATA = "${localstatedir}/rootdirs/userdata"
+
+#/var/lib/edge
+EDGE_STATE = "${localstatedir}/lib/edge"
+
+#/opt
+EDGE_OPT = "${base_prefix}/opt"
+
+#/var/run/edge 
+EDGE_RUN = "${localstatedir}/run/edge"
+
+#TEMPORARY: The following lines keep the system as it was in 2.3 
+EDGE_BIN = "/wigwag/system/bin"
+EDGE_LIB = "/wigwag/system/lib"
+EDGE_ETC = "/wigwag/system/etc"
+EDGE_DATA = "/userdata"
+EDGE_STATE = "/wigwag/system/var/lib"
+EDGE_OPT = "/wigwag/system/opt"
+#END TEMPORARY
+
+EDGE_CNI_CONF = "${EDGE_ETC}/cni/net.d"
+EDGE_CNI_BIN = "${EDGE_OPT}/cni/bin"
+EDGE_KUBELET_STATE = "${EDGE_STATE}/kubelet"
+EDGE_COREDNS_STATE = "${EDGE_STATE}/coredns"
+#-----------------------------------------------------------------------------------------------------#
+#                                              _       _     _                                        #
+#                             __   ____ _ _ __(_) __ _| |__ | | ___  ___                              #
+#                             \ \ / / _` | '__| |/ _` | '_ \| |/ _ \/ __|                             #
+#                              \ V / (_| | |  | | (_| | |_) | |  __/\__ \                             #
+#                               \_/ \__,_|_|  |_|\__,_|_.__/|_|\___||___/                             #
+#                                                                                                     #
+#-----------------------------------------------------------------------------------------------------#
+EDGE_NODEIP = "10.240.0.1"
+EDGE_PODCIDR = "10.240.0.0/24"
+EDGE_NODEDNSPORT = "53"
+
+
+
+#*Note: comments examplifying paths assume a standard version of yocto with an unmodified bitbake.conf
+
+edge_replace_vars() {
+	for file in "$@"; do
+	    sed -i  "s/EDGE_BIN/$(echo ${EDGE_BIN} | sed -e 's/\\/\\\\/g; s/\//\\\//g; s/&/\\\&/g')/g" "$file"
```

Comment:
What does this do?

Reply:
Similar concept to a macro in c.  Replaces constant text with other fixed text from the edge.bbclass file.  Why?  To enable customers to change the installation location for Pelion programs by modifying one file.  Use case example:  Currently we put all executables in /Pelion/system/bin.  This is done in many recipes.  e.g. If all Pelion recipes inherit edge.bbclass, and use the variable "${EDGE_BIN}" the management of where edge executables are stored can be centralized to edge.bbclass.  Moreover, runtime scripts, configuration files, etc can also use these fixed constants (e.g. EDGE_BIN) by calling the function edge_replace_vars() in right before the do_install() script of a recipe. 
examples of this can be found in coredns_1.8.3.bb, kubelet_git.bb,   kube-router_git.bb


example summary:
including edge: https://github.com/PelionIoT/meta-pelion-edge/pull/311/files#diff-49df459969129d53e1b29714b32c6ea24b1d7c93e9e361c9eed24e98fff3670cR5

Using a variable from edge.bbclass (in local recipe): https://github.com/PelionIoT/meta-pelion-edge/pull/311/files#diff-49df459969129d53e1b29714b32c6ea24b1d7c93e9e361c9eed24e98fff3670cR55

Using a variable from edge.bbclass (in a file installed): https://github.com/PelionIoT/meta-pelion-edge/pull/311/files#diff-b34bce18f9268029f98e6c10e1b1dcb3e88b214d0c26f86025c4a7d29fa5cdb6R81

Processing a file installed by the recipe: https://github.com/PelionIoT/meta-pelion-edge/pull/311/files#diff-49df459969129d53e1b29714b32c6ea24b1d7c93e9e361c9eed24e98fff3670cR48

Reply:
Thanks for the summary. I'd recommend documenting this well, since it can be a bit confusing if encountering the file with no context.  The mechanism is quite good, at least I can't think of a better solution off the top of my head for centralized variable substitution. 

Perhaps it could be considered if the variables could be soft assignments. That way user wouldn't have to modify  recipes for configuration, instead they could define the configuration in local.conf. I'm not 100% sure if this is possible since I can't remember if classes can read read local.conf or have soft assignments, but it could improve the usability a bit if it'd work.